app/src/main/java/com/samco/trackandgraph/antlr/ast/Validation.kt

```diff
@@ -0,0 +1,70 @@
+//   Copyright 2021 Federico Tomassetti
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+
+//  Modified
+
+
+package com.samco.trackandgraph.antlr.ast
+
+import com.samco.trackandgraph.antlr.evaluation.DoubleDeclarationError
+import com.samco.trackandgraph.antlr.evaluation.NotDeclaredError
+import com.samco.trackandgraph.antlr.evaluation.ReferencedBeforeDeclarationError
+import com.samco.trackandgraph.antlr.evaluation.UnexistingVariableError
+import java.util.*
+
+
+open class Error(message: String, var position: Position) : Exception(message) {
```

Comment:
This performs checks on the code while parsing. I adapted the original code a bit to allow external variables. Otherwise not much to change